packages/vite/src/node/optimizer/esbuildDepPlugin.ts

```diff
@@ -190,30 +194,32 @@ export function esbuildDepPlugin(
         }
       })
 
-      build.onLoad(
-        { filter: /.*/, namespace: 'browser-external' },
-        ({ path: id }) => {
-          return {
-            contents:
-              `export default new Proxy({}, {
-  get() {
-    throw new Error('Module "${id}" has been externalized for ` +
-              `browser compatibility and cannot be accessed in client code.')
-  }
-})`
-          }
-        }
-      )
+      // Returning empty contents that will be turned by ESBuild
+      // into a CommonJS module exporting an empty object.
+      // This is what ESBuild does when bundling a dependency
+      // starting from a non-browser-external entrypoint.
+      build.onLoad({ filter: /.*/, namespace: 'browser-external' }, () => {
+        return { contents: '' }
+      })
```

Comment:
Would this PR work without this proxy change? I don't think it's a good idea to remove the friendly warnings for users who don't uses yarn pnp

Reply:
Only partially.

This problem https://github.com/vitejs/vite/pull/6493#discussion_r786515167 will remain.

>  I don't think it's a good idea to remove the friendly warnings for users who don't uses yarn pnp

It's already is not there for most cases. Because without PnP ESBuild handles it on its own and does not call this onLoad.

Here https://github.com/vitejs/vite/pull/6493#discussion_r803460203 is comparison of output with and without PnP.

Reply:
Interesting. I've seen the proxy's error out in the wild before though, but that was a few months ago. Perhaps across esbuild updates that it changed 🤔 Might have to double-check this myself when I get the time.

Reply:
That'd be great.

I've updated my test repo with more import patterns (I think they now cover all the code paths through which a browser-externalized module can be generated) and instructions on running them with patched/unpatched Vite and node-modules/pnp mode.

https://github.com/swandir/vite-browser-external-pnp

Hope that helps.

Reply:
Hey @swandir. I checked the repro today, really appreciate the detailed descriptions!

> > I don't think it's a good idea to remove the friendly warnings for users who don't uses yarn pnp
> 
> It's already is not there for most cases. Because without PnP ESBuild handles it on its own and does not call this onLoad.

Thats seems to be so because you have `browser: { "./node": false }`. If it weren't there, the proxy would kick in and report the messages, and I think this would be helpful for end-users who accidentally import node packages. So I think we should still keep the proxy in place.

But I know you mention that this causes errors like `[ERROR] No matching export in "browser-external:./lib/node/util" for import "promisify"`. I was under the assumption that the nested proxy hack would fix this error, and it's indeed an issue outside of pnp too. Let's take tackle this in a separate PR then, and focus on pnp here for now.

Besides reverting the proxy changes, the rest looks good to me!

Reply:
> Thats seems to be so because you have browser: { "./node": false }. If it weren't there, the proxy would kick in and report the messages

`./node` would have been bundled in without the proxy. The proxy kicks in only if a module is a builtin or is disabled via `browser` field.

Honestly, I think it's a bit better without this proxy module, there are three possible outcomes depending on an import location: (1) you get a `export default` proxy from resolvePlugin.load, or (2) a CommonJS proxy here, or (3) an empty CommonJS module from ESBuild. And without PnP (2) is the most common anyway.

Practically speaking, anything you import (default or named) from a module disabled for the browser is `undefined`. Most of the time you either access a property on an imported symbol or call it as a function, which both fail at runtime without the proxy too.

Sure, "fs is externalized for browser compatibility" is nicer than "Cannot read properties of undefined (reading 'fs')". But I'm not sure if it's worth additional complexity and edge cases it brings.

Anyway, I reverted the CommonJS proxy and added a comment in the code.
It does the trick for PnP right now, and everything else is kind of out of scope and indeed can be handles separately.

Thanks!

Reply:
I actually meant that we revert back to the proxy as ESM (without the cjs hack), but if you prefer to leave it here, I might have to check that locally do verify again that all is fine 😅 

> Sure, "fs is externalized for browser compatibility" is nicer than "Cannot read properties of undefined (reading 'fs')". But I'm not sure if it's worth additional complexity and edge cases it brings.

I think the goal is here is to have the friendly message work 90% of the time. We don't need things to be correct to handle all the edge cases, but just enough that most people benefit from it.

Reply:
> I think the goal is here is to have the friendly message work 90% of the time. We don't need things to be correct to handle all the edge cases, but just enough that most people benefit from it.

But the moment someone hits the case where things break, they do want them to work correctly 😅

> I actually meant that we revert back to the proxy as ESM (without the cjs hack) but if you prefer to leave it here, I might have to check that locally do verify again that all is fine 

Sadly, that'll leave Vite failing to start for any project that use PnP with a double node/browser dependency that happen to have a named import from a node-only module. (And we happen to be using such a dependency).

I'd really like to have the fix upstream in some form, so that I don't have to maintain a patch.

But as far as I can see, there's no way around changing this proxy for that. As it breaks PnP compatibility specifically, because without PnP this proxy module is never generated for what seems to be the most common case — using a package with browser mapping.